Update README.md (#34)
NCSU-High-Powered-Rocketry-Club/AirbrakesV2: README.md

> # Air BrakesV2 🚀

Reviewer: ```suggestion
# AirbrakesV2 🚀
```
nah one word is better

> ## Overview

Reviewer: we need a ToC here

## Overview
This project controls the extension of airbrakes using a servo motor. It includes modules for logging data and managing the servo motor.

### Installation

#### Install uv:

> [uv](https://docs.astral.sh/uv/getting-started/installation/) is a python and project manager, which makes handling different python versions and virtual enviornments easier.

Reviewer: no `uv` :(

Author: I thought about it more -- while it's a nice thing, I don't think it's required/related specifically to our project. If someone wants to use it, that's great, but I don't think we should make them do it -- just trying to simplify installation

> This project is for controlling our Air brakes system with the goal of making our rocket "hit" its target apogee. We have a Raspberry Pi 4 as the brains of our system which runs our code. It connects to a servo motor to control the extension of our air brakes and an [IMU](https://www.microstrain.com/inertial-sensors/3dm-cx5-25) (basically an altimeter, accelerometer, and gyroscope). The code follows the [finite state machine](https://www.tutorialspoint.com/design_pattern/state_pattern.htm) design pattern, using the [`AirbrakesContext`](https://github.com/NCSU-High-Powered-Rocketry-Club/AirbrakesV2/blob/main/airbrakes/airbrakes.py) to manage interactions between the states, hardware, logging, and data processing. 

Reviewer: ```suggestion
This project is for controlling our Air brakes system with the goal of making our rocket "hit" its target apogee. We have a Raspberry Pi 4 as the brains of our system which runs our code. It connects to a servo motor to control the extension of our air brakes and an [IMU](https://www.microstrain.com/inertial-sensors/3dm-cx5-15) (basically an altimeter, accelerometer, and gyroscope). The code follows the [finite state machine](https://www.tutorialspoint.com/design_pattern/state_pattern.htm) design pattern, using the [`AirbrakesContext`](https://github.com/NCSU-High-Powered-Rocketry-Club/AirbrakesV2/blob/main/airbrakes/airbrakes.py) to manage interactions between the states, hardware, logging, and data processing. 
```

We have cx-15. Might be worth updating the class docstring of `IMU` as well.

> https://github.com/user-attachments/assets/0c72a9eb-0b15-4fbf-9e62-f6a69e5fadaa

Reviewer: nice, I thought this would be too sensitive / confidential information lol

Author: Ehh I don't love the idea of keeping secrets. If someone directly tells us to close things down, we will, but until then I like keeping it open.

1. **Gets data from the IMU**
2. **Processes the data** in the Data Processor (calculates speed, averages, maximums, etc.)
3. **Updates the current state** with the processed data
4. **Controls the servo extension** based on the current state's instructions (e.g., extends air brakes to slow down the rocket)
5. **Logs all data** from the IMU, Data Processor, Servo, and States
```mermaid
flowchart TD
    A[Main Loop] --> B[Airbrakes Context]
    B --> C[IMU - Get Data] --> B
    B --> D[Data Processor - Process Data] --> B
    B --> F[Update Current State] --> B
    B --> G[Servo - Control Air Brakes] --> B

>     B --> E[Logger - Log Data]

Reviewer: this is fine but I was picturing a more detailed diagram, showing the 2 child processes (IMU and logger) and then also including the other classes, like `IMUDataPacket`, `ProcessedDataPacket`, and `LoggedDataPacket`, `States`, etc. I can sketch a diagram during today's meeting and ask an llm to convert it to mermaid

```

### Launch Data

This is our interest launch flight data, altitude over time. The different colors of the line are different states the rocket goes through:
1. Stand By - when the rocket is on the rail on the ground
2. Motor Burn - when the motor is burning and the rocket is accelerating
3. Coast - when the motor has burned out and the rocket is coasting, this is when air brakes will be deployed
4. Free Fall - when the rocket is falling back to the ground after apogee, this is when the air brakes will be
retracted
5. Landed - when the rocket has landed on the ground
<img alt="graph" src="https://github.com/user-attachments/assets/39cf0556-d388-458b-8668-64177506c9de" width="70%">

### File Structure

> We have put great care into keeping the file structure of this project organized and concise. Try to be intentional with where placing new files or directories.

Reviewer: ```suggestion
We have put great effort into keeping the file structure of this project organized and concise. Try to be intentional with placing new files or directories.
```

├── README.md
```

## Local Setup

**This project uses Python 3.12. Using an older version may not work since we use newer language features**

### Clone the repository:

```
git clone https://github.com/NCSU-High-Powered-Rocketry-Club/AirbrakesV2.git
cd AirbrakesV2
```

### Set up a virtual environment:

```bash
python -m venv .venv

# For Linux
source .venv/bin/activate
# For Windows

> venv\Scripts\activate

Reviewer: ```suggestion
.\.venv\Scripts\activate
```
isn't it this?

```

### Install the required dependencies:

```bash
pip install .[dev]
```

> _There are libraries that only fully work when running on the Pi (gpiozero, mscl), so if you're having trouble importing them locally, program the best you can and test your changes on the pi._

Reviewer: we should mention `mscl` installation instructions.

Author: I was thinking about that, but do you think it's necessary? It's already installed/setup on the pi, and having it installed on your laptop doesn't matter that much

Reviewer: It's important for the future people who want to setup `mscl` on their Pi's, e.g. I had to do it when swapping the SD card

```bash
python3 main.py m
```
### Running Tests
Our CI pipeline uses [pytest](https://pytest.org) to run tests. You can run the tests locally to ensure that your changes are working as expected.

To run the tests, run this command from the project root directory:
```bash
pytest
```
If you make a change to the code, please make sure to update or add the necessary tests.

### Running the Linter

Our CI also tries to maintain code quality by running a linter. We use [Ruff](https://docs.astral.sh/ruff/).

To run the linter, and fix any issues it finds, run:
```bash
ruff check . --fix --unsafe-fixes
```
To format the code, run:
```bash
ruff format .
```

## Pi Usage

### Connecting to the Pi (SSH)
In order to connect to the Pi, you need first to set up a mobile hotspot with the name `HPRC`, password `tacholycos`, and `2.4 GHz` band. Next, turn on the Pi and it will automatically connect to your hotspot. Once it's connected, find the Pi's IP Address, and in your terminal run:
```bash
ssh pi@[IP.ADDRESS]
# Its password is raspberry
cd AirbrakesV2/
```

### Install and start the pigpio daemon on the Raspberry Pi:

> _Every time the pi boots up, you must run this in order for the servo to work. We have already added this command to run on startup, but you may want to confirm that it is running, e.g. by using `htop`._

Reviewer: ```suggestion
_Every time the pi boots up, you must run this in order for the servo to work. We have already added this command to run on startup, but you may want to confirm that it is running, e.g. by using [`htop`](https://htop.dev/)._
```
